### ISSUE TITLE
[KEYBOARD]: Form review edit button focus

### ISSUE THREAD

**Issue by Mottie** [department-of-veterans-affairs/va.gov-team#4613]
## User Story

As a veteran, I need to be able to use a keyboard to navigate through the form review in a logical manner if I don't have access to a mouse.

## Description

This issue was originally brought up with a Higher-Level Review a11y review ([ref](https://github.com/department-of-veterans-affairs/va.gov-team/issues/3566#issuecomment-559214317)), but is an issue associated with all forms. cc @jenstrickland 

The problem is with the review page accordion focus handling. After activating the edit button, focus remains on that button with no indication to the user where they are in the form or that they need to press tab to get to the first focusable element.

## Acceptance Criteria

- [ ] A meaningful heading should be available for each editable section
- [ ] The meaningful heading should update text when users press the Edit button. Ideally we would change a heading like "Contact information" to "Edit contact information".
- [ ] Focus should move to the heading element when users press the Edit button, and when they submit their edit-in-place form. See this [Edit in place focus demo Codepen](https://codepen.io/tpierce_402/pen/NWNerLm) for correct behavior. 
- [ ] Focus should be verifiable by console.log when user presses Edit, and when user submits their form<br /><br />
  ```javascript
  document.activeElement
  ```
- [ ] The heading should not "jump" around visually when users enter the edit-in-place mode. We want to change the text, present the form, and not introduce jank, as much as possible.

## Steps to Recreate

1. Complete any form and get to the review and submit page
2. Navigate or click on any accordion header and expand it while maintaining focus (yellow outline).

    <img width="594" alt="form review expanded accordion" src="https://user-images.githubusercontent.com/57469/69751647-8f0dee00-111d-11ea-9aa6-e46eea39344a.png" />

3. Press <kbd>Tab</kbd> to move focus to the "Edit" button.
4. Press <kbd>Enter</kbd> to switch to edit mode.
5. At this point, the first problem is encountered. The focus remains on the now hidden "Edit" button with no message stating that the area is interactive

    <img width="723" alt="Screen Shot 2019-11-27 at 1 55 46 PM" src="https://user-images.githubusercontent.com/57469/69751698-a77e0880-111d-11ea-8efa-14ee2c229546.png">

6. Pressing <kbd>Tab</kbd> again, now moves the focus to the first focusable elements inside the expanded accordion, but no indication that the element or area is interactive

    <img width="616" alt="Screen Shot 2019-11-27 at 1 56 40 PM" src="https://user-images.githubusercontent.com/57469/69751757-c11f5000-111d-11ea-9996-8cfcb6d73abe.png">

## Possible Fixes (optional)

* Move focus from the edit button to the first focusable element. Here is some untested code that should accomplish this:

```js
export function focusOnFirstElement(block, options) {
  // List from https://html.spec.whatwg.org/dev/dom.html#interactive-content
  const focusableElements = [
    '[href]',
    'button',
    'details',
    'input:not([type="hidden"])',
    'label',
    'select',
    'textarea',
    '[tabindex]:not([tabindex="-1"])', // focusable, but not tabbable
    // focusable elements not used in our form system
    // 'audio[controls]',
    // 'embed',
    // 'iframe',
    // 'img[usemap]',
    // 'object[usemap]',
    // 'video[controls]',
  ];
  block
    ?.querySelectorAll(focusableElements.join(','))
    ?.filter(el => !el.disabled)
    ?.[0]
    ?.focus(options);
}

// Use it as follows:
const button = document.activeElement;
const reviewBlock = button.closest('form')?.querySelector('.review');
focusOnFirstElement(reviewBlock);
```

**Comment by bkjohnson** [comment]
> There is no indicator that the input is editable

Do we want to add a `<span>` with `aria-live="polite"` that says something like "You may now edit these fields" at the top of the section?

**Comment by jenstrickland** [comment]
If we move focus to the first editable field it should read the label and input, which provides the cue it is now editable.

**Comment by meganhkelley** [comment]
Relevant Slack thread: https://dsva.slack.com/archives/CQH357ZTP/p1597240608005300

**Comment by meganhkelley** [comment]
@caw310 handing this off to your team

**Comment by caw310** [comment]
@SebVortex , The designer, @GnatalieH on our team is looking into this.

**Comment by SebVortex** [comment]
@GnatalieH 
It's been a while. Has there been any progress on the design work?

**Comment by GnatalieH** [comment]
@SebVortex I'm not a developer so I'm not sure why this is assigned to me. I did communicate the proper behavior to @bkjohnson a while back. @caw310 can you make sure a developer on our team is assigned this fix? Thanks.

**Comment by GnatalieH** [comment]
> If we move focus to the first editable field it should read the label and input, which provides the cue it is now editable.

This should be the fix.

**Comment by Mottie** [comment]
I believe this fix has already been implemented. I just checked on a form locally, and activating the edit button moves the focus to the content inside the page, and activating "update page" button moves the focus back to the edit button.

**Comment by GnatalieH** [comment]
> I believe this fix has already been implemented. I just checked on a form locally, and activating the edit button moves the focus to the content inside the page, and activating "update page" button moves the focus back to the edit button.

Thank you for checking on this. When focus moves from the edit button to the first editable field, are the label and input read so that it provides a cue the item is editable?

**Comment by Mottie** [comment]
The legend or label that is focused it read... I didn't check to see if it indicated that the item was editable.

**Comment by SebVortex** [comment]
@Mottie Well if it can be confirmed that it's working, then I can close this issue and it should free up some others.

**Comment by GnatalieH** [comment]
> @Mottie Well if it can be confirmed that it's working, then I can close this issue and it should free up some others.

I think it can be closed. As long as the label and input are read, that should indicate the item is editable (according to an older comment). Thanks!

**Comment by SebVortex** [comment]
Alright, will do. Thanks everyone!

**Comment by SebVortex** [comment]
Alright, will do. Thanks everyone!

**Comment by SebVortex** [comment]
Alright, will do. Thanks everyone!

**Comment by mrblanco1** [comment]
Hello @GnatalieH - Thank you for all your help! We really appreciate it. Per Sebastian (SebVortex) and your comments this issue can be closed now.  Do you mind closing it since we don't have permissions and add a detailed comment please?

Thank you in advanced!

Abe Blanco - AFS Comparison Tool Delivery Manager

**Comment by GnatalieH** [comment]
Closing this issue because it has been resolved. Thanks, everyone!
